src/rebar_git_resource.erl

```diff
@@ -188,6 +188,55 @@ maybe_warn_local_url(Url) ->
         _ -> ok
     end.
 
+git_cache_dir(Url, Tag) ->
+    case os:getenv("REBAR_GIT_CACHE_DIR") of
+        false ->
+            nocache;
+        Dir ->
+            Full = case parse_git_url_as_cache_path(Url) of
+                nocache ->
+                    nocache;
+                "/" ++ Path ->
+                    filename:join([Dir, Path, Tag])
+            end,
+            DotGit = filename:join(Full, ".git"),
+            case filelib:is_dir(DotGit) of
+                true ->
+                    {hit, Full};
+                false ->
+                    {miss, Full}
+            end
+    end.
+
+parse_git_url_as_cache_path(Url) ->
+    case parse_git_url(Url) of
+        {ok, {Host, Path}} ->
+            filename:join([Host, Path]);
+        Other ->
+            ?DEBUG("Failed to parse url ~ts ~p", [Url, Other]),
+            nocache
+    end.
+
+git_clone2(tag, GitVsn, Url, Dir, Tag) ->
+    CloneF = fun() -> git_clone(tag, GitVsn, Url, Dir, Tag) end,
+    case git_cache_dir(Url, Tag) of
+        nocache ->
+            ok = CloneF();
+        {hit, CacheDir} ->
+            %% cache hit, copy from the cache
+            ?DEBUG("Git cache hit: ~ts", [CacheDir]),
+            rebar_file_utils:cp_r([CacheDir], Dir);
```

Comment:
We may be getting inconsistent state of repositories due to race conditions.
For example:
- one runner checked there no cache, started clone operation
- another comes online 2 seconds later, checks there is `.git` in place, decides cache is in place starts copying - but the files are only partially there

May be add some kind of lock file? If it's in place, some process has already started populating the cache.
And may be another file to indicate if the cache is consistent and can be used.

Reply:
changed back to use `--reference`